### PR TITLE
Cost optimization works

### PR DESCRIPTION
**Pull request by 123MwanjeMike**
**_WHAT DOES THIS PR DO?_**
- [update k8s playbook hosts](https://github.com/airqo-platform/AirQo-api/pull/1271/commits/c40ef0dfc97519e6aff81ae6a5fd392bf81e6b1f)
- [update dependency installation playbook](https://github.com/airqo-platform/AirQo-api/pull/1271/commits/a3ae77af884ba146eacadfff39e73882f2339f9e)
- [add all worker nodes](https://github.com/airqo-platform/AirQo-api/pull/1271/commits/fc8a750cfef79ab6163124b017bd1c2c408fb180)
- Create an [hourly_k8s_runners instance schedule](https://github.com/airqo-platform/AirQo-api/pull/1271/commits/56c1e8b177bd7d8eb6774bae9561c00e23733fd4)
- Adds preemptive k8s worker nodes for cronjobs

**_WHAT ISSUES ARE RELATED TO THIS PR?_**
- [PLATOPS-16](https://airqoteam.atlassian.net/browse/PLATOPS-16?atlOrigin=eyJpIjoiNmM1ZjE5Y2UyMTkzNGI4Y2FiOWNiMGMwODQ5YTNjMDUiLCJwIjoiaiJ9)

**_HOW DO I TEST OUT THIS PR?_**
- n/a

**_ARE THERE ANY RELATED PRs?_**
- Forked from #1242 

[PLATOPS-16]: https://airqoteam.atlassian.net/browse/PLATOPS-16?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: infra/terraform/gcp/modules/ComputeResourcePolicy/hourly-k8s-runners.tf [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    vm_start_schedule {
+      schedule = "55 * * * *"
+    }
+    vm_stop_schedule {
+      schedule = "10 * * * *"
+    }
+    time_zone = "Africa/Kampala"
+  }
+}
+# terraform import google_compute_resource_policy.hourly_k8s_runners projects/${var.project_id}/regions/${var.region}/resourcePolicies/hourly-k8s-runners
```
Comment (123MwanjeMike) [inline infra/terraform/gcp/modules/ComputeResourcePolicy/hourly-k8s-runners.tf:16]: [hourly_k8s_runners instance schedule](https://github.com/airqo-platform/AirQo-api/pull/1271/commits/56c1e8b177bd7d8eb6774bae9561c00e23733fd4)

File: infra/ansible/playbooks/k8s/k8s-hosts-prod [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+          preemptive_node_name: airqo-k8s-worker-0
+    k8s_workers:
+      hosts:
+        10.240.0.21
+        10.240.0.69
+        10.240.0.104
+      children:
+        preemptive_node:
+          hosts:
+            10.240.0.20
```
Comment (123MwanjeMike) [inline infra/ansible/playbooks/k8s/k8s-hosts-prod:15]: Prod inventory

File: infra/ansible/playbooks/k8s/k8s-hosts-stage [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        10.240.0.100:
+          preemptive_node_name: airqo-stage-k8s-worker-0
+    k8s_workers:
+      hosts:
+        10.240.0.102
+        10.240.0.70
+      children:
+        preemptive_node:
+          hosts:
+            10.240.0.36:
```
Comment (123MwanjeMike) [inline infra/ansible/playbooks/k8s/k8s-hosts-stage:14]: Stage inventory

File: infra/terraform/gcp/modules/ComputeInstance/airqo-stage-k8s-worker-0.tf [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  metadata = {
+    startup-script = "sudo ufw allow ssh"
+  }
+
+  network_interface {
+    access_config {
+      network_tier = "PREMIUM"
+    }
+    network    = "airqo-k8s-cluster"
```
Comment (123MwanjeMike) [inline infra/terraform/gcp/modules/ComputeInstance/airqo-stage-k8s-worker-0.tf:22]: stage preemptive vm

File: infra/terraform/gcp/modules/ComputeInstance/airqo-k8s-worker-0.tf [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    auto_delete = true
+    source      = "https://www.googleapis.com/compute/v1/projects/airqo-250220/zones/europe-west1-b/disks/airqo-k8s-worker-0"
+  }
+
+  metadata = {
+    startup-script = "sudo ufw allow ssh"
+  }
+
+  network_interface {
+    access_config {
```
Comment (123MwanjeMike) [inline infra/terraform/gcp/modules/ComputeInstance/airqo-k8s-worker-0.tf:18]: prod preemptive vm

File: infra/ansible/setup/etc/hosts [hunk 6]
```diff
@@ -1,19 +1,16 @@
...
+      hosts:
+        mongos-prod-1.europe-west1-b.c.airqo-250220.internal
+      children:
+        cfgsvrs_prod:
+          hosts:
+            cfgsvr-prod-[1:3].europe-west1-b.c.airqo-250220.internal
+        shards_prod:
+          hosts:
+            shard-prod-[1:3].europe-west1-b.c.airqo-250220.internal
+
```
Comment (123MwanjeMike) [inline infra/ansible/setup/etc/hosts]: [change all inventories to yaml style](https://github.com/airqo-platform/AirQo-api/pull/1271/commits/21e5afa2bbb60db4c588c0eb66f1760d831978f2)

File: infra/ansible/playbooks/k8s/templates/remove_preemptive_node.sh.j2 [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+kubectl cordon {{ preemptive_node_name }} --kubeconfig=/etc/kubernetes/admin.conf
+kubectl delete node {{ preemptive_node_name }} --kubeconfig=/etc/kubernetes/admin.conf
```
Comment (123MwanjeMike) [inline infra/ansible/playbooks/k8s/templates/remove_preemptive_node.sh.j2:3]: Remove preemptive bash script

File: infra/ansible/playbooks/k8s/templates/leave_k8s_cluster.sh.j2 [hunk 8]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+kubeadm reset -f
```
Comment (123MwanjeMike) [inline infra/ansible/playbooks/k8s/templates/leave_k8s_cluster.sh.j2:2]: Leave k8s cluster script

File: infra/ansible/playbooks/k8s/templates/join_k8s_cluster.sh.j2 [hunk 9]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+{{ hostvars[groups['k8s_master'].0].join_command }} > node_joined.txt
```
Comment (123MwanjeMike) [inline infra/ansible/playbooks/k8s/templates/join_k8s_cluster.sh.j2:2]: Join k8s cluster script

File: infra/ansible/playbooks/k8s/preemptive-worker.yml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    - name: Create cronjob to run the remove_preemptive_node script at minute 8 of every hour
+      cron:
+        name: "Remove preemptive k8s node from cluster"
+        minute: "8"
+        job: "/bin/bash /root/remove_preemptive_node.sh"
+
+    - name: Get join command with a token that expires in 1 hour
+      shell: kubeadm token create --print-join-command --ttl 1h
+      register: join_command_raw
```
Comment (123MwanjeMike) [inline infra/ansible/playbooks/k8s/preemptive-worker.yml:19]: Preemptive worker playbook

**Review by Baalmart — COMMENTED** [review]
Hi @123MwanjeMike , could we have a workshop (working meeting) on this?

Also, any quick resources for members to get up to speed before that working meeting?

Thanks.

**Review by Baalmart — COMMENTED** [review]
Thanks @123MwanjeMike for this.

In order to ensure that most are onboarded to make future reviews easier?

Are we able to have more REVIEW comments added to this PR?

Thanks!

**Comment by 123MwanjeMike** [comment]
Thank you @Baalmart 
Yes, a workshop does sound good indeed.
In regard to review comments, I had added some for all the new resource types added.

File: infra/ansible/playbooks/k8s/k8s-hosts-prod [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+        10.240.0.11
+    k8s_workers:
+      hosts:
+        10.240.0.21
+        10.240.0.69
+        10.240.0.104
+      children:
+        preemptive_node:
+          hosts:
+            10.240.0.20
```
Comment (NoahNsimbe) [inline infra/ansible/playbooks/k8s/k8s-hosts-prod:14]: Hi @123MwanjeMike is it possible to store these ip address in environment variables
e.g https://www.digitalocean.com/community/tutorials/how-to-create-a-kubernetes-cluster-using-kubeadm-on-ubuntu-16-04

Comment (123MwanjeMike) [reply infra/ansible/playbooks/k8s/k8s-hosts-prod:14]: Hi @NoahNsimbe 
Please provide more context on why we would need to do that.
Note that this is an inventory file using prod k8s hosts' internal IPs for the VPC network.

Comment (NoahNsimbe) [reply infra/ansible/playbooks/k8s/k8s-hosts-prod:14]: thanks @123MwanjeMike for the clarification. I hadn't noticed that its an inventory file.

Comment (123MwanjeMike) [reply infra/ansible/playbooks/k8s/k8s-hosts-prod:14]: No worries @NoahNsimbe 
I shifted from .ini style to .yaml format due to ease of readability

**Pull request merged**